service/backend/pchain/construction.go

```diff
@@ -0,0 +1,337 @@
+package pchain
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/vms/components/avax"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/coinbase/rosetta-sdk-go/types"
+
+	"github.com/ava-labs/avalanche-rosetta/mapper"
+	pmapper "github.com/ava-labs/avalanche-rosetta/mapper/pchain"
+	"github.com/ava-labs/avalanche-rosetta/service"
+	"github.com/ava-labs/avalanche-rosetta/service/backend/common"
+)
+
+var (
+	errUnknownTxType = errors.New("unknown tx type")
+	errUndecodableTx = errors.New("undecodable transaction")
+	errNoTxGiven     = errors.New("no transaction was given")
+)
+
+// ConstructionDerive implements /construction/derive endpoint for P-chain
+func (b *Backend) ConstructionDerive(ctx context.Context, req *types.ConstructionDeriveRequest) (*types.ConstructionDeriveResponse, *types.Error) {
+	return common.DeriveBech32Address(b.fac, mapper.PChainNetworkIdentifier, req)
+}
+
+// ConstructionPreprocess implements /construction/preprocess endpoint for P-chain
+func (b *Backend) ConstructionPreprocess(
+	ctx context.Context,
+	req *types.ConstructionPreprocessRequest,
+) (*types.ConstructionPreprocessResponse, *types.Error) {
+	matches, err := common.MatchOperations(req.Operations)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInvalidInput, err)
+	}
+
+	reqMetadata := req.Metadata
+	if reqMetadata == nil {
+		reqMetadata = make(map[string]interface{})
+	}
+	reqMetadata[pmapper.MetadataOpType] = matches[0].Operations[0].Type
+
+	return &types.ConstructionPreprocessResponse{
+		Options: reqMetadata,
+	}, nil
+}
+
+// ConstructionMetadata implements /construction/metadata endpoint for P-chain
+func (b *Backend) ConstructionMetadata(
+	ctx context.Context,
+	req *types.ConstructionMetadataRequest,
+) (*types.ConstructionMetadataResponse, *types.Error) {
+	opMetadata, err := pmapper.ParseOpMetadata(req.Options)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInvalidInput, err)
+	}
+
+	var suggestedFee *types.Amount
+	var metadata *pmapper.Metadata
+	switch opMetadata.Type {
+	case pmapper.OpImportAvax:
+		metadata, suggestedFee, err = b.buildImportMetadata(ctx, req.Options)
+	case pmapper.OpExportAvax:
+		metadata, suggestedFee, err = b.buildExportMetadata(ctx, req.Options)
+	case pmapper.OpAddValidator, pmapper.OpAddDelegator:
+		metadata, suggestedFee, err = b.buildStakingMetadata(req.Options)
```

Comment:
@bilgehansahin are these the only tx Rosetta can build? Why?

Reply:
We have added the transactions necessary for staking AVAX. We left the remaining transactions for future updates.